packages/gguf/src/gguf.ts

```diff
@@ -0,0 +1,275 @@
+		new Uint8Array(this.buffer).set(buf, this.chunk * HTTP_CHUNK_SIZE);
+		this.chunk += 1;
+	}
+	/**
+	 * Check whether we need to fetch a new chunk
+	 */
+	async check(offset: number) {
+		if (this.view.byteLength - offset < HTTP_DATA_LEEWAY) {
+			await this.fetchChunk();
+		}
+	}
+}
+
+function readVersionedSize(view: DataView, byteOffset: number, version: Version): bigint {
+	switch (version) {
+		case 1: {
+			const n = view.getUint32(byteOffset, true);
+			return BigInt(n);
+		}
+		case 2:
+		case 3: {
+			return view.getBigUint64(byteOffset, true);
+		}
+	}
+}
+
+function readString(view: DataView, offset: number): { value: string; newOffset: number } {
+	const length = view.getBigUint64(offset, true);
+	const value = new TextDecoder().decode(view.buffer.slice(offset + 8, offset + 8 + Number(length)));
+	return { value, newOffset: offset + 8 + Number(length) };
+}
+
+function readMetadataValue(
+	view: DataView,
+	type: GGUFValueType,
+	offset: number
+): { value: MetadataValue; newOffset: number } {
+	switch (type) {
+		case GGUFValueType.UINT8:
+			return { value: view.getUint8(offset), newOffset: offset + 1 };
+		case GGUFValueType.INT8:
+			return { value: view.getInt8(offset), newOffset: offset + 1 };
+		case GGUFValueType.UINT16:
+			return { value: view.getUint16(offset, true), newOffset: offset + 2 };
+		case GGUFValueType.INT16:
+			return { value: view.getInt16(offset, true), newOffset: offset + 2 };
+		case GGUFValueType.UINT32:
+			return { value: view.getUint32(offset, true), newOffset: offset + 4 };
+		case GGUFValueType.INT32:
+			return { value: view.getInt32(offset, true), newOffset: offset + 4 };
+		case GGUFValueType.FLOAT32:
+			return { value: view.getFloat32(offset, true), newOffset: offset + 4 };
+		case GGUFValueType.BOOL:
+			return { value: view.getUint8(offset) !== 0, newOffset: offset + 1 };
+		case GGUFValueType.STRING:
+			return readString(view, offset);
+		case GGUFValueType.ARRAY: {
+			const arrayType = view.getUint32(offset, true);
+			const arrayLength = view.getBigUint64(offset + 4, true);
+			let arrayOffset = offset + 12;
+			const arrayValues: MetadataValue[] = [];
+			for (let i = 0; i < arrayLength; i++) {
+				const { value, newOffset } = readMetadataValue(view, arrayType, arrayOffset);
+				arrayValues.push(value);
+				arrayOffset = newOffset;
+			}
+			return { value: arrayValues, newOffset: arrayOffset };
+		}
+		case GGUFValueType.UINT64:
+			return { value: view.getBigUint64(offset, true), newOffset: offset + 8 };
+		case GGUFValueType.INT64:
+			return { value: view.getBigInt64(offset, true), newOffset: offset + 8 };
+		case GGUFValueType.FLOAT64:
+			return { value: view.getFloat64(offset, true), newOffset: offset + 8 };
+	}
+}
+
+export type GGUFMetadata = {
+	version: Version;
+	tensor_count: bigint;
+	kv_count: bigint;
+} & Record<string, MetadataValue>;
+
+export interface GGUFTensorInfo {
+	name: string;
+	n_dims: number;
+	shape: bigint[];
+	type: GGMLQuantizationType;
```

Comment:
```suggestion
	dtype: GGMLQuantizationType;
```

to be consistent with general ml lingo

Reply:
9366d4ae27a0ff6e289b7f63280ff56d0c80a715